---
title: "Assignment 2 - Differential Gene expression and Preliminary ORA"
output:
  html_document:
    df_print: paged
---

Install necessary packages.

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))    
  install.packages("BiocManager")
if (!requireNamespace("GEOmetadb", quietly = TRUE))    
  BiocManager::install("GEOmetadb")
if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")
if (!requireNamespace("ggplot2", quietly = TRUE))
  install.packages("ggplot2")
if (!requireNamespace("reshape2", quietly = TRUE))
  BiocManager::install("reshape2")
if (!requireNamespace("HGNChelper", quietly = TRUE))
  BiocManager::install("HGNChelper")
if (!requireNamespace("limma", quietly = TRUE))
  BiocManager::install("limma")
if (!requireNamespace("knitr", quietly = TRUE))
  BiocManager::install("knitr")
if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
  install.packages("ComplexHeatmap")
if (!requireNamespace("circlize", quietly = TRUE))
  install.packages("circlize")
if (!requireNamespace("Biobase", quietly = TRUE))
  BiocManager::install("Biobase")
if (!requireNamespace("MEAL", quietly = TRUE))
  BiocManager::install("MEAL")
```

## Select an Expression Data Set.

Get the data for my chosen GEO id. I chose to work with the data set from an experiment that uses RNA-sequencing to study stem cell derived cortical excitatory (cExN) and cortical inhibitory (cIN) neurons from three individuals with varying affectation of autism syndrome disorder (ASD) (Lewis, Meganathan and Baldridge 2019).
```{r}
myGEOID <- 'GSE129806'
suppFiles = GEOquery::getGEOSuppFiles(myGEOID)
```

Read the data in. There are 2 files provided one with the RPKM data and one with the raw counts data. I will work with the second file which has the counts data.
```{r}
fileNames = rownames(suppFiles)
data <- read.delim(fileNames[2], check.names=FALSE, header=TRUE)
```

Check what the columns represent. They represent the cExN or cIN samples from the 4 individuals. "UC" prefix represents unaffected control group individual. "AP" prefix represents affected proband individual. "IS" prefix represents the intermediated phenotype sister. "UM" prefix respesents unaffected mother.
```{r}
colnames(data)
```

## Clean the data and map to HUGO symbols
Rename column "name" to "gene_id" for more clarity.
```{r}
colnames(data)[1] <- 'gene_id'
```

Remove any NAs.
```{r}
na.omit(data)
```

See how many genes we have. 
```{r}
dim(data)
```

Take a look at the genes in the data set and check for duplicates.
```{r}
sort(table(data$gene_id), decreasing = TRUE)[1:25]
```
Looking at the results, there are no duplicates.
However, looks like there are gene symbols mistakenly converted to date by Excel. I will remove these dates as it's unclear what genes they were converted from.
```{r}
data <- data[!grepl('^[0-9]{1,2}[-][a-zA-Z]{3}$', data$gene_id),]
```

Check the row names to see what they are. Rename each row to the gene id.
```{r}
rownames(data)[1:10]
rownames(data) <- data$gene_id
```

Check if the symbols are all HGNC or if they are up to date using HGNCHelper.
```{r}
replacements <- HGNChelper::checkGeneSymbols(data$gene_id)
# Update the gene id with the newest HGNC symbols if available
data$gene_id <- ifelse(!replacements$Approved & !is.na(replacements$Suggested.Symbol), replacements$Suggested.Symbol, replacements$x)
```

See how many of the remaining gene_ids could not be mapped to a HGNC symbol. Some of these include
long non-coding RNAs and pseudogenes.
```{r}
not_mapped <- length(which(!HGNChelper::checkGeneSymbols(data$gene_id)$Approved))
```

In the experiment, there were at least 3 or more biological replicates for each subject. I will use n=3 to perform removal of features without at least 1 read per million in 3 of the samples. Use edgeR package to calculate CPM and remove low counts.
```{r}
cpms = edgeR::cpm(data[,3:34])
rownames(cpms) <- data[,1]

keep = rowSums(cpms > 1) >= 3
data_filtered = data[keep,]
rows_removed = dim(data)[1]-dim(data_filtered)[1]
```

Check the number of genes now.
```{r}
dim(data_filtered)
```

Create a grouping from the data set.
```{r}
# Based on code from lecture 4
samples <- data.frame(lapply(colnames(data)[3:34], FUN=function(x){unlist(strsplit(x, split="_"))[c(1,2)]}))
colnames(samples) = colnames(data)[3:34]
rownames(samples) = c("patients", "cell_type")
samples <- data.frame(t(samples))
```

## Apply Normalization

I will apply TMM to the data set. I chose to use this method because most of the genes were not differentially expressed and that the RNA seq data does not to be modified. 
```{r}
# Based on code from lecture 4
data_filtered_matrix <- as.matrix(data_filtered[,3:34])
rownames(data_filtered_matrix) <- data_filtered$gene_id
d = edgeR::DGEList(counts=data_filtered_matrix, group=samples$cell_type)
```

Calculate the normalization factor and get it.
```{r}
# Based on code from lecture 4
d = edgeR::calcNormFactors(d)
normalized_counts <- edgeR::cpm(d)
```

Modify the regular data set for plotting.
```{r}
data_to_plot <- reshape2::melt((edgeR::cpm(data_filtered[,3:34], log=TRUE)))
colnames(data_to_plot) <- c("gene_id", "sample", "cpm")
```

Modify the normalized data set for plotting.
```{r}
normalized_data_to_plot <- reshape2::melt(log2(normalized_counts))
colnames(normalized_data_to_plot) <- c("gene_id", "sample", "cpm")
```

Create a pre-normalized boxplot to visualize the data.
```{r}
ggplot2::ggplot(data_to_plot, ggplot2::aes(x=sample, y=cpm)) +
  ggplot2::geom_boxplot() + 
  ggplot2::theme(axis.text.x=ggplot2::element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::labs(title="Original cIN and cExN RNASeq Samples", y="log2-CPM")
```

Create a normalized boxplot to visualize the data.
```{r}
ggplot2::ggplot(normalized_data_to_plot, ggplot2::aes(x=sample, y=cpm)) +
  ggplot2::geom_boxplot() + 
  ggplot2::theme(axis.text.x=ggplot2::element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggplot2::labs(title="Normalized cIN and cExN RNASeq Samples", y="log2-CPM")
```

Create a pre-normalized density plot to visualize the data.
```{r}
ggplot2::ggplot(data_to_plot, ggplot2::aes(x=cpm, color=sample)) +
  ggplot2::geom_density() +
  ggplot2::labs(title="Original cIN and cExN RNASeq Samples", y="density of log2-CPM", x="")
```

Create a normalized density plot to visualize the data.
```{r}
ggplot2::ggplot(normalized_data_to_plot, ggplot2::aes(x=cpm, color=sample)) +
  ggplot2::geom_density() +
  ggplot2::labs(title="Normalized cIN and cExN RNASeq Samples", y="density of log2-CPM", x="")
```

The density graph of the data set looks like it follows a bimodal distribution.

## Intepret, and Document

### What are the control and test conditions of the dataset?
Looking at the data set and the paper, the test condition is the cortical excitatory (cExN) and cortical inhibitory (cIN) neurons from the three first-degree relatives with absent, intermediate and severe expressivity of autism syndrome disorder (ASD). The control condition is the cExN and cIN neurons from the unrelated individual. The samples of the three first-degree relatives are prefixed with UM (Unaffected mother), IS (Intermediated Sister) and AP (Affected proband). The sample of the unrelated individual is prefixed with UC (Unrelated, unaffected control).

### Why is the dataset of interest to you?
This data set is interesting to me because I am currently working in a lab that does research related to autism spectrum disorder.

### Were there expression values that were not unique for specific genes? How did you handle these?
There weren't multiple expression values that mapped to a single gene.

### Were there expression values that could not be mapped to current HUGO symbols?
Yes there were `r not_mapped` expression values that could not be
mapped to a current HUGO symbol.

### How many outliers were removed?
`r rows_removed` outliers were removed.

### How did you handle replicates?
Excluding the control cell line, there were 3 biological replicates from each of the two clonal lines for each test subject. For now, I decided to keep the replicates separate and analyze them independently.

### What is the final coverage of your dataset?
There was approximately 30 million reads per sample according to the paper (Lewis, Meganathan and Baldridge 2019)

## Differential Gene Expression
The purpose of this section is to conduct differential expression analysis and rank genes with the normalized expression set from assignment 1.

### Summary of Assignment 1
In assignment 1, I downloaded the data set of interest. The data set is from an experiment that uses RNA-sequencing to study stem cell derived cortical excitatory (cExN) and cortical inhibitory (cIN) neurons from three individuals with varying affectation of autism syndrome disorder (ASD) (Lewis, Meganathan and Baldridge 2019). To explore the data, the null values, mistakes (dates instead of genes) and outliers were removed. The data was then normalized using TMM normalization. The pre-normalization and post-normalization data was graphed out using boxplots and density curves.

### Plot MDS
From the MDS plot we can see two things. One is the significant dissimilarity between the two cell types. In dark green we have the cortical inhibitory neuron cells and in the blue we have the cortical excitatory neuron cells. As well we can see the that both cell types of the affected proband (AP) are clustered near the top. Both cell types of the mother (UM) and sibling (IS) of the affected proband are clustered near the middle and the unaffected control cell types are clustered near the bottom. Since we're interested in seeing which genes may be attributed to autism syndrome disorder, we'll focus our analysis on which patient (UM, IS, AP or UC) the cells come from.
```{r}
limma::plotMDS(d, labels=rownames(samples), col = c("darkgreen", "blue")[factor(samples$cell_type)])
```
### Loading normalized data and producing heat map
```{r}
# Take a look at the normalized data from the previous assignment
knitr::kable(normalized_counts[1:5, 1:5], type="html")
```
```{r}
# Convert it to a number matrix
heatmap_matrix <- as.matrix(normalized_counts)
# Scale and centre each row around the mean
heatmap_matrix <- t(scale(t(heatmap_matrix)))

# Create the heat map
if(min(heatmap_matrix) == 0){
  heatmap_col = circlize::colorRamp2(c(0, max(heatmap_matrix)),
                                      c("white", "red"))
} else {
  heatmap_col = circlize::colorRamp2(c(min(heatmap_matrix), 0,
                                       max(heatmap_matrix)), c("blue", "white", "red"))
}
current_heatmap <- ComplexHeatmap::Heatmap(as.matrix(heatmap_matrix),
                                           show_row_dend = TRUE, show_column_dend = TRUE,
                                           col=heatmap_col, show_column_names = TRUE, show_row_names = FALSE,
                                           show_heatmap_legend = TRUE)
current_heatmap
```

Now we want to plot the MDS using limma R package. We can tell there are 3 distinct clusters. The cluster to the top left show the cIN neuron cells which all have similar logFC dim1 and dim2 scores. The cluster to the top right shows the affected proband, unaffected mother and intermediate sibling cExN cells, which are distinct from the unaffected control cExN cells. This is useful because we want to see analyze the gene expression in cells from patients with autism syndrome disorder or their relatives.
```{r}
limma::plotMDS(heatmap_matrix, col = rep(c("darkgreen", "blue"), 10))
```


To see the data in a different perspective, we can colour by patient and do another MDS plot. We can see that the clustering for patients is better in the cIN cells than the cExN cells. Once again we see that there is significant difference between the gene expression of UC cExN cells and the gene expression of the other 3 patient cells.
```{r}
pat_colors <- rainbow(4)
pat_colors <- unlist(lapply(pat_colors,FUN=function(x){rep(x,4)}))
limma::plotMDS(heatmap_matrix, col = pat_colors)
```

For the next step, we want to create a model matrix and fit the data we have into it. We will focus on using the patient column for analysis.

```{r}
# Set model to patients
model_design <- model.matrix(~ samples$patients)
knitr::kable(model_design, type="html")
```


After, we can construct a expression matrix and minimal set from the normalized counts.
```{r}
# There was an error citing duplicate row names. We will use make.names to make them unique.
rownames(normalized_counts) = make.names(rownames(normalized_counts), unique=TRUE)
# Calculate expression matrix and minimal set
expression_matrix <- as.matrix(normalized_counts)
minimal_set <- Biobase::ExpressionSet(assayData=expression_matrix)
```

Then fit the data to the model and apply empirical Bayes to calculate P values.
```{r}
fit <- limma::lmFit(minimal_set, model_design)
fit2 <- limma::eBayes(fit,trend=TRUE)

topfit <- limma::topTable(fit2,
                          coef=ncol(model_design),
                          adjust.method = "BH",
                          number = nrow(expression_matrix))
output_hits <- merge(rownames(normalized_counts),
                     topfit,
                     by.y=0, by.x=1,
                     all.y=TRUE)
output_hits <- output_hits[order(output_hits$P.Value),]
# Take a look at the top 10 hits
knitr::kable(output_hits[1:10,], type="html")
```

From the results we can see that 2499 genes had a P value < 0.05.
```{r}
length(which(output_hits$P.Value < 0.05))
```

Once corrected, that number goes down to 714 genes.
```{r}
simple_adj_p_values <- length(which(output_hits$adj.P.Val < 0.05))
simple_adj_p_values
```

Let's graph the p-values to see if we have any potential problems. The graph shows that we have a set of good p-values.
Now we can apply multiple hypothesis test correction

```{r}
ggplot2::ggplot(output_hits, ggplot2::aes(x=P.Value)) +
  ggplot2::geom_histogram() +
  ggplot2::labs(title="P values", y="counts", x="p-value")
```

### Multiple Hypothesis Correction
We now want to improve our results and account for the variability seen in different patients and cell types. We'll create a design matrix that accounts for both cell type and patient.
```{r}
model_design_ct <- model.matrix(~ samples$cell_type + samples$patients)
knitr::kable(model_design_ct, type="html")
```

Now we repeat the previous step of fitting the data to our new model.
```{r}
fit_ct <- limma::lmFit(minimal_set, model_design_ct)
fit_ct2 <- limma::eBayes(fit_ct,trend=TRUE)

topfit_ct <- limma::topTable(fit_ct2,
                          coef=ncol(model_design_ct),
                          adjust.method = "BH",
                          number = nrow(expression_matrix))
output_hits_ct <- merge(rownames(normalized_counts),
                     topfit_ct,
                     by.y=0, by.x=1,
                     all.y=TRUE)
output_hits_ct <- output_hits_ct[order(output_hits_ct$P.Value),]
# Take a look at the top 10 hits
knitr::kable(output_hits_ct[1:10,], type="html")

```

Again take a look at the p-values.
From the results we can see that 5682 genes had a P value < 0.05.
```{r}
length(which(output_hits_ct$P.Value < 0.05))
```

Once corrected, that number goes down to 3171 genes.
```{r}
complex_adj_p_values <- length(which(output_hits_ct$adj.P.Val < 0.05))
complex_adj_p_values
```

### Model selection
Now I want to compare the two models I have.

```{r}
simple_model_pvalues <- data.frame(gene_id = output_hits$x, simple_pvalue = output_hits$P.Value)

complex_model_pvalues <- data.frame(gene_id = output_hits_ct$x, complex_pvalue = output_hits_ct$P.Value)

two_models_pvalues <- merge(simple_model_pvalues, complex_model_pvalues, by.x=1, by.y=1)

two_models_pvalues$colour <- "black"
two_models_pvalues$colour[two_models_pvalues$simple_pvalue<0.05] <- "orange"
two_models_pvalues$colour[two_models_pvalues$complex_pvalue<0.05] <- "blue"
two_models_pvalues$colour[two_models_pvalues$simple_pvalue<0.05 &  two_models_pvalues$complex_pvalue<0.05] <- "red"


plot(two_models_pvalues$simple_pvalue, two_models_pvalues$complex_pvalue, col = two_models_pvalues$colour, xlab = "simple model p-values", ylab ="Complex model p-values", main="Simple vs Complex Limma")
legend("topleft", inset=0.1, legend=c("Simple", "Complex", "Both", "Not Significant"), 
       fill = c("orange", "blue", "red", "black"))
```

In the above plot, we can see that the simple model which considers patients only has a better fit to the p-values of both models.


We want to include the FMR1 gene in the plot which is an important gene for autism syndrome disorder.

```{r}
two_models_pvalues$colour <- "grey"
two_models_pvalues$colour[two_models_pvalues$gene_id == "FMR1"] <- "red"
plot(two_models_pvalues$simple_pvalue, two_models_pvalues$complex_pvalue, col = two_models_pvalues$colour, xlab = "simple model p-values", ylab ="Complex model p-values", main="Simple vs Complex Limma")
points(two_models_pvalues[which(two_models_pvalues$gene_id == "CHCHD2"), 2:3],pch=20, col="red", cex=1.5 )
legend("topleft", inset=0.1, legend=c("FMR1", "rest"), 
       fill = c("red", "grey"), cex=0.7)
```
To see differentially expressed genes, we use a MA plot.
```{r}
volcano_plot <- cbind(output_hits$logFC, -log10(output_hits$adj.P.Val))
colnames(volcano_plot) <- c("logFC", "-log10 of adj. P value")
point.col <- ifelse(output_hits$logFC < 0, "red", "blue")
plot(volcano_plot, pch = 8, col = point.col, cex = 0.5, main = "logFC vs adjusted p value Graph ")
```
The volcano graph above shows the genes that are upregulated in blue and downregulated in red. It tells us the distribution of the upregulated vs downregulated.


#### Question 1:
There were `r simple_adj_p_values` genes that were differentially expressed for the simple model considering patient type. There were `r complex_adj_p_values` genes that were differentially expressed for complex model considering both patien type and cell type. I used the p-value of 0.05 since that is the most common p-value widely accepted.

#### Question 2:
I chose to use the Benjamini-Hochberg method because it is straightforward and intuitive to use. As well since there are over 20,000 genes to consider, the false discovery rate can cause many false positives. The Benjamini-Hochberg test helps with reducing false positive rate, it in turns help reeduces the number of false positives in my model.

## References
* Lewis, E.M.A., Meganathan, K., Baldridge, D. et al. Cellular and molecular characterization of multiplex autism in human induced pluripotent stem cell-derived neurons. Molecular Autism 10, 51 (2019). https://doi.org/10.1186/s13229-019-0306-0
* Robinson, M.D., Oshlack, A. A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biol 11, R25 (2010). https://doi.org/10.1186/gb-2010-11-3-r25
* Martin Morgan (2019). BiocManager: Access the Bioconductor Project Package Repository. R package version 1.30.10. https://CRAN.R-project.org/package=BiocManager
* Zhu Y, Davis S, Stephens R, Meltzer PS, Chen Y. GEOmetadb: powerful alternative search engine for the Gene Expression Omnibus. Bioinformatics. 2008 Dec 1;24(23):2798-800. doi:10.1093/bioinformatics/btn520. Epub 2008 Oct 7. PubMed PMID: 18842599; PubMed Central PMCID:
  PMC2639278.
* Levi Waldron and Markus Riester (2019). HGNChelper: Identify and Correct Invalid HGNC Human Gene Symbols and MGI Mouse Gene Symbols. R package version 0.8.1. https://github.com/waldronlab/HGNChelper
* H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.
* Hadley Wickham (2007). Reshaping Data with the reshape Package. Journal of Statistical Software, 21(12), 1-20. URL http://www.jstatsoft.org/v21/i12/.
* Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics 26, 139-140
* McCarthy DJ, Chen Y and Smyth GK (2012). Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation. Nucleic Acids Research 40, 4288-4297
